File: packages/button-android/src/main/kotlin/com/porto/ds/button/PSButtonViewModel.kt
package com.porto.ds.button

import androidx.compose.ui.graphics.Color
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class PSButtonViewModel : ViewModel() {
    private val _uiState = MutableStateFlow(ButtonUiState())
    val uiState: StateFlow<ButtonUiState> = _uiState.asStateFlow()

    fun processIntent(
        intent: ButtonIntent,
        onPress: (() -> Unit)? = null,
    ) {
        val current = _uiState.value
        when (intent) {
            is ButtonIntent.Press -> {
                if (!current.isDisabled && !current.isLoading) onPress?.invoke()
            }
            is ButtonIntent.HoverEnter -> {
                if (!current.isDisabled && !current.isLoading) {
                    updateInteraction(ButtonInteractionState.Hover)
                }
            }
            is ButtonIntent.HoverExit -> {
                if (!current.isDisabled && !current.isLoading) {
                    updateInteraction(ButtonInteractionState.Default)
                }
            }
            is ButtonIntent.Configure -> {
                val impliedDisabled = intent.isDisabled || intent.isLoading
                val newState =
                    if (impliedDisabled) {
                        ButtonInteractionState.Disabled
                    } else {
                        ButtonInteractionState.Default
                    }
                _uiState.value =
                    current.copy(
                        variant = intent.variant,
                        size = intent.size,
                        isDisabled = impliedDisabled,
                        isLoading = intent.isLoading,
                        interactionState = newState,
                        tokenSet = resolveTokenSet(intent.variant, newState),
                    )
            }
        }
    }

    fun resolveTokenSet(
        variant: ButtonVariant,
        state: ButtonInteractionState,
    ): ButtonTokenSet {
        val bg = PSButtonTokens.Colors.Background
        val fg = PSButtonTokens.Colors.Foreground
        val brd = PSButtonTokens.Colors.Border
        val isDisabled = state == ButtonInteractionState.Disabled || state == ButtonInteractionState.Loading

        val backgroundColor: Color =
            when {
                variant is ButtonVariant.Primary && state == ButtonInteractionState.Hover -> bg.PrimaryHover
                variant is ButtonVariant.Primary && state == ButtonInteractionState.Active -> bg.PrimaryActive
                variant is ButtonVariant.Primary && isDisabled -> bg.PrimaryDisabled
                variant is ButtonVariant.Primary -> bg.Primary
                variant is ButtonVariant.Secondary && state == ButtonInteractionState.Hover -> bg.SecondaryHover
                variant is ButtonVariant.Secondary && state == ButtonInteractionState.Active -> bg.SecondaryActive
                variant is ButtonVariant.Secondary -> bg.Secondary
                variant is ButtonVariant.Ghost && state == ButtonInteractionState.Hover -> bg.GhostHover
                variant is ButtonVariant.Ghost -> bg.Ghost
                variant is ButtonVariant.Destructive && state == ButtonInteractionState.Hover -> bg.DestructiveHover
                variant is ButtonVariant.Destructive && isDisabled -> bg.DestructiveDisabled
                variant is ButtonVariant.Destructive -> bg.Destructive
                else -> bg.Primary
            }

        val foregroundColor: Color =
            when {
                isDisabled -> fg.PrimaryDisabled
                variant is ButtonVariant.Primary -> fg.Primary
                variant is ButtonVariant.Secondary -> fg.Secondary
                variant is ButtonVariant.Ghost -> fg.Ghost
                variant is ButtonVariant.Destructive -> fg.Destructive
                else -> fg.Primary
            }

        val (borderColor, borderWidth) =
            if (variant is ButtonVariant.Secondary) {
                (if (isDisabled) brd.SecondaryDisabled else brd.Secondary) to PSButtonTokens.BorderWidth.value
            } else {
                Color.Transparent to 0f
            }

        return ButtonTokenSet(
            backgroundColor = backgroundColor,
            foregroundColor = foregroundColor,
            borderColor = borderColor,
            borderWidth = borderWidth,
            focusRingColor = brd.FocusRing,
        )
    }

    private fun updateInteraction(newState: ButtonInteractionState) {
        val current = _uiState.value
        _uiState.value =
            current.copy(
                interactionState = newState,
                tokenSet = resolveTokenSet(current.variant, newState),
            )
    }
}

File: packages/button-ios/Sources/Button/PSButtonViewModel.swift
import Foundation
import Combine

public final class PSButtonViewModel: ObservableObject {
    @Published public private(set) var state: PSButtonState = .default
    @Published public private(set) var tokenSet: PSButtonTokenSet

    private let variant: PSButtonVariant
    private let size: PSButtonSize

    public init(variant: PSButtonVariant = .primary, size: PSButtonSize = .md) {
        self.variant = variant
        self.size = size
        self.tokenSet = PSButtonViewModel.resolveTokenSet(variant: variant, state: .default)
    }

    public func handlePress(isDisabled: Bool, isLoading: Bool, action: () -> Void) {
        guard !isDisabled && !isLoading else { return }
        state = .active
        action()
        state = .default
    }

    public static func resolveTokenSet(variant: PSButtonVariant, state: PSButtonState) -> PSButtonTokenSet {
        let t = PSButtonTokens.self
        let isDisabled = state == .disabled || state == .loading

        let bg: UIColor
        let fg: UIColor
        let border: UIColor
        let borderWidth: CGFloat

        switch (variant, state) {
        case (.primary, .hover):   bg = t.Color.Background.primaryHover
        case (.primary, .active):  bg = t.Color.Background.primaryActive
        case (.primary, .disabled), (.primary, .loading): bg = t.Color.Background.primaryDisabled
        case (.primary, _):        bg = t.Color.Background.primary
        case (.secondary, .hover): bg = t.Color.Background.secondaryHover
        case (.secondary, .active): bg = t.Color.Background.secondaryActive
        case (.secondary, _):      bg = t.Color.Background.secondary
        case (.ghost, .hover):     bg = t.Color.Background.ghostHover
        case (.ghost, .active):    bg = t.Color.Background.ghostHover
        case (.ghost, _):          bg = t.Color.Background.ghost
        case (.destructive, .hover): bg = t.Color.Background.destructiveHover
        case (.destructive, .disabled), (.destructive, .loading): bg = t.Color.Background.destructiveDisabled
        case (.destructive, _):    bg = t.Color.Background.destructive
        }

        fg = isDisabled ? t.Color.Foreground.primaryDisabled : {
            switch variant {
            case .primary:     return t.Color.Foreground.primary
            case .secondary:   return t.Color.Foreground.secondary
            case .ghost:       return t.Color.Foreground.ghost
            case .destructive: return t.Color.Foreground.destructive
            }
        }()

        if variant == .secondary {
            border = isDisabled ? t.Color.Border.secondaryDisabled : t.Color.Border.secondary
            borderWidth = t.borderWidth
        } else {
            border = .clear
            borderWidth = 0
        }

        return PSButtonTokenSet(
            backgroundColor: bg,
            foregroundColor: fg,
            borderColor: border,
            borderWidth: borderWidth,
            focusRingColor: t.Color.Border.focusRing
        )
    }
}
